File: src_python/shogi_api/board.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import os, sys, copy

from .move import get_legal_moves
from .util import *
from .piece import Piece
                


class Board:

    def __init__(self, all_pieces=None):

        if all_pieces is None:

            loc_piece_dict = generate_initial_loc_piece_dict()
            pieces_in_hand = {piece_name : 0 for piece_name in PieceName_Hand}

            self.all_pieces = {
                "main" : loc_piece_dict,
                "hand" : {"sente" : copy.deepcopy(pieces_in_hand), "gote" : copy.deepcopy(pieces_in_hand)}
                }

        else:
            self.all_pieces = all_pieces

        self.is_sente = True  # 手番
        self.board_history = []  # 盤面の状態を毎回保存

        self.legal_moves = get_legal_moves(copy.deepcopy(self))  # 可能な指し手をだす
        self.board_history.append(copy.deepcopy(self))  # historyに追加

    def __str__(self):
        
        return self.kif_str()

    def kif_str(self):

        kif_str = ""

        for row_num in range(1, 10):
            for col_num in range(1, 10)[::-1]:
                
                loc_str = str(col_num) + str(row_num)
                piece = self.all_pieces["main"][loc_str]

                if piece is None:
                    the_str = "n一"

                else:
                    piece_name = piece.name
                    is_sente = piece.is_sente
                    piece_name_kanji = PieceName_normal2kanji[piece_name]
                    if is_sente:
                        the_str = "a{}".format(piece_name_kanji)
                    else:
                        the_str = "v{}".format(piece_name_kanji)

                kif_str += the_str
            
            if row_num != 9:
                kif_str += "\n"

        return kif_str

    def move(self, move):
        
        '''
        '''

        print("~"*20)
        print(self.all_pieces["main"]["99"].loc)
        
        assert (len(move) == 4) or (len(move) == 5)

        loc_from = move[:2]
        loc_to = move[2:4]

        # その手が可能かチェック
        assert move in self.legal_moves

        # 手駒を打つ
        if loc_from in PieceName_Hand:
            
            piece_name = loc_from
            
            # 手駒から減らす
            if self.is_sente:
                self.all_pieces["hand"]["sente"][piece_name] -= 1
            else:
                self.all_pieces["hand"]["gote"][piece_name] -= 1

            # 駒オブジェクトを生成して盤上に配置
            the_piece = Piece(name=piece_name, is_sente=self.is_sente)
            self.all_pieces["main"][loc_to] = the_piece


        # 駒を移動させる
        else:
            
            piece = self.all_pieces["main"][loc_from]
            piece.loc = loc_to  # 位置を移動

            # 成る時
            if move[-1] == "+":
                piece.name = PieceName_before2after[piece.name]

            # 駒をとる時
            piece_there = self.all_pieces["main"][loc_to]
            if piece_there is not None:
                assert piece_there.is_sente != self.is_sente
                # もし成り駒を取るときは名前を元に戻す
                if piece_there.name in PieceName_Nari:
                    piece_there.name = PieceName_after2before[piece_there.name]
                # 持ち駒に加える
                if self.is_sente:
                    self.all_pieces["hand"]["sente"][piece_there.name] += 1
                else:
                    self.all_pieces["hand"]["gote"][piece_there.name] += 1

            # 位置を移動
            self.all_pieces["main"][loc_from] = None
            self.all_pieces["main"][loc_to] = copy.deepcopy(piece)

        print("%"*20)
        print(self.all_pieces["main"]["99"].loc)

        # 手番を交代
        self.is_sente = not self.is_sente
        # legal_movesを出しておく
        self.legal_moves = get_legal_moves(copy.deepcopy(self))

        print("e"*20)
        print(self.all_pieces["main"]["99"].loc)

        # historyに追加
        # self.board_history.append(copy.deepcopy(self))

    
    def get_past(self, num):
        
        '''
        num手目の盤面を返す
        '''
        
        return self.board_history[num]
    

File: app/static/board.js



class Board{

    constructor (all_pieces=null, is_sente=null, board_history=null, legal_moves=null){

        if (all_pieces == null){

            let loc_piece_dict = generate_initial_loc_piece_dict();
            let pieces_in_hand_sente = {};
            let pieces_in_hand_gote = {};

            for (let piece_name of PieceName_Hand){
                pieces_in_hand_sente[piece_name] = 0;
                pieces_in_hand_gote[piece_name] = 0;
            }

            this.all_pieces = {
                "main" : loc_piece_dict,
                "hand" : {"sente" : pieces_in_hand_sente, "gote" : pieces_in_hand_gote}
            }

            this.is_sente = true;  // 手番
            this.board_history = [];  // 盤面の状態を毎回保存
            this.legal_moves = get_legal_moves(deepcopy_Board(this));  // 可能な指し手をだす
            this.board_history.push(deepcopy_Board(this));  // historyに追加
        }

        else{
            this.all_pieces = all_pieces;
            this.is_sente = is_sente;  // 手番
            this.board_history = board_history;  // 盤面の状態を毎回保存
            this.legal_moves = legal_moves;  // 可能な指し手をだす
        }

    }



    // def kif_str(this):

    //     kif_str = ""

    //     for row_num in range(1, 10):
    //         for col_num in range(1, 10)[::-1]:
                
    //             loc_str = str(col_num) + str(row_num)
    //             piece = this.all_pieces["main"][loc_str]

    //             if piece is None:
    //                 the_str = "n一"

    //             else:
    //                 piece_name = piece.name
    //                 is_sente = piece.is_sente
    //                 piece_name_kanji = PieceName_normal2kanji[piece_name]
    //                 if is_sente:
    //                     the_str = "a{}".format(piece_name_kanji)
    //                 else:
    //                     the_str = "v{}".format(piece_name_kanji)

    //             kif_str += the_str
            
    //         if row_num != 9:
    //             kif_str += "\n"

    //     return kif_str

    move(move){
        
        if ((move.length != 4) && (move.length != 5)){
            window.alert("kora");
            exit;
        }

        let loc_from = move.slice(0, 2);
        let loc_to = move.slice(2, 4);

        // その手が可能かチェック
        if (this.legal_moves.indexOf(move) < 0){
            window.alert("kora");
            exit;
        }

        // 手駒を打つ
        if (PieceName_Hand.indexOf(loc_from) >= 0){
            
            let piece_name = loc_from
            
            // 手駒から減らす
            if (this.is_sente){
                this.all_pieces["hand"]["sente"][piece_name] -= 1;
            }
            else{
                this.all_pieces["hand"]["gote"][piece_name] -= 1;
            }

            // 駒オブジェクトを生成して盤上に配置
            let the_piece = new Piece(piece_name, this.is_sente, loc_to);
            this.all_pieces["main"][loc_to] = the_piece;
        }


        // 駒を移動させる
        else{
            
            let piece = deepcopy_Piece(this.all_pieces["main"][loc_from]);
            piece.loc = loc_to;  // 位置を移動

            // 成る時
            if (move[move.length - 1] == "+"){
                piece.name = PieceName_before2after[piece.name];
            }

            // 駒をとる時
            let piece_there = this.all_pieces["main"][loc_to];

            if (piece_there != null){

                if (piece_there.is_sente == this.is_sente){
                    window.alert("kora");
                    exit;
                }


                // もし成り駒を取るときは名前を元に戻す
                if (PieceName_Nari.indexOf(piece_there.name) >= 0){
                    piece_there.name = PieceName_after2before[piece_there.name];
                }


                // 持ち駒に加える
                if (this.is_sente){
                    this.all_pieces["hand"]["sente"][piece_there.name] += 1;
                }
                else{
                    this.all_pieces["hand"]["gote"][piece_there.name] += 1;
                }

            }

            // 位置を移動
            this.all_pieces["main"][loc_from] = null;
            this.all_pieces["main"][loc_to] = piece;

        }


        // 手番を交代
        this.is_sente = !this.is_sente;

        // legal_movesを出しておく
        this.legal_moves = get_legal_moves(deepcopy_Board(this));


        // historyに追加
        // this.board_history.append(copy.deepcopy(this))

    }

    
    get_past(num){
        
        // num手目の盤面を返す
        
        return deepcopy_Board(this.board_history[num]);
    }
}


function deepcopy_Board(board){

    if (board.is_sente){var is_sente_copy = true;}else{var is_sente_copy=false;}
    var all_pieces_copy = deepcopy_all_pieces(board.all_pieces);
    var board_history_copy = $.extend(true, [], board.board_history);
    var legal_moves_copy = $.extend(true, [], board.legal_moves);

    let new_Board = new Board(all_pieces_copy, is_sente_copy, board_history_copy, legal_moves_copy);
    
    return new_Board;
}

function deepcopy_all_pieces(all_pieces){
    
    let main_copy = {};
    for (let key of Object.keys(all_pieces["main"])){
        let the_piece = all_pieces["main"][key];
        if (the_piece == null){
            main_copy[key] = null;
        }
        else{
            main_copy[key] = deepcopy_Piece(the_piece);
        }
    }
    // let main_copy = $.extend(true, {}, all_pieces["main"]);
    let sente_copy = $.extend(true, {}, all_pieces["hand"]["sente"]);
    let gote_copy = $.extend(true, {}, all_pieces["hand"]["gote"]);
    let hand_copy = {"sente" : sente_copy, "gote" : gote_copy};

    return {"main" : main_copy, "hand" : hand_copy};
}